src/Microsoft.AspNetCore.Blazor.Browser/Interop/JavaScriptInvoke.cs

```diff
@@ -0,0 +1,172 @@
+        private static readonly IDictionary<string, object> ResolvedFunctions = new Dictionary<string, object>();
+
+        private const string InvokePromiseCallback = "invokePromiseCallback";
+
+        public static string FindDotNetMethod(string methodOptions)
+        {
+            var result = ResolvedFunctionRegistrations.GetOrAdd(methodOptions, opts =>
+            {
+                var options = JsonUtil.Deserialize<MethodOptions>(methodOptions);
+                var argumentDeserializer = GetOrCreateArgumentDeserializer(options);
+                var invoker = GetOrCreateInvoker(options, argumentDeserializer);
+
+                var invokerRegistration = NextFunction.ToString();
+                NextFunction++;
+                if (ResolvedFunctions.ContainsKey(invokerRegistration))
+                {
+                    throw new InvalidOperationException($"A function with registration '{invokerRegistration}' was already registered");
+                }
+                else
+                {
+                    ResolvedFunctions[invokerRegistration] = invoker;
+                }
+
+                return invokerRegistration;
+            });
+
+            return result;
+        }
+
+        public static string InvokeDotNetMethod(string registration, string callbackId, string methodArguments)
+        {
+            // We invoke the dotnet method and wrap either the result or the exception produced by
+            // an error into an invocation result type. This invocation result is just a discriminated
+            // union with either success or failure.
+            try
+            {
+                return InvocationResult<object>.Success(InvokeDotNetMethodCore(registration, callbackId, methodArguments));
+            }
+            catch (Exception e)
+            {
+                var exception = e;
+                while (exception.InnerException != null)
+                {
+                    exception = exception.InnerException;
+                }
+
+                return InvocationResult<object>.Fail(exception);
+            }
+        }
+
+        private static object InvokeDotNetMethodCore(string registration, string callbackId, string methodArguments)
+        {
+            if (!ResolvedFunctions.TryGetValue(registration, out var registeredFunction))
+            {
+                throw new InvalidOperationException($"No method exists with registration number '{registration}'.");
+            }
+
+            if (!(registeredFunction is Func<string, object> invoker))
+            {
+                throw new InvalidOperationException($"The registered invoker has the wrong signature.");
+            }
+
+            var result = invoker(methodArguments);
+            if (callbackId != null && !(result is Task))
+            {
+                var methodSpec = ResolvedFunctionRegistrations.Single(kvp => kvp.Value == registration);
```

Comment:
Should this be something other than a `.Single`? isn't this a dictionary?

Reply:
I'm not worried about the exception cases. The goal here is to provide a good error message, not to be efficient. If you run into this, you clearly have a bug in your code.